File: packages/Admob/ios/Classes/FLTTeadsAdapterSettings.swift
//
//  FLTTeadsAdapterSettings.swift
//  teads_admob_adapter
//
//  Created by Thibaud Saint-Etienne on 17/10/2022.
//

import Flutter
import TeadsSDK
import GoogleMobileAds

public class FLTTeadsAdapterSettings: NSObject, FlutterPlugin {

    static let instance = FLTTeadsAdapterSettings()
    
    var adapterSettings: TeadsAdapterSettings = {
        let settings = TeadsAdapterSettings()
        settings.addExtras(TeadsAdPlacementSettings.pluginFlutter, for: TeadsAdPlacementSettings.pluginKey)
        // Fluter version is always 1.0.0
        settings.addExtras("1.0.0", for: TeadsAdPlacementSettings.pluginVersionKey)
        return settings
    }()
    
    static var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(name: "teads_admob_adapter/teads_adapter_settings", binaryMessenger: registrar.messenger())
        if let channel = channel {
            registrar.addMethodCallDelegate(instance, channel: channel)
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "disableCrashMonitoring":
            adapterSettings.disableCrashMonitoring()
            result(try? adapterSettings.asDictionary())
        case "disableTeadsAudioSessionManagement":
            adapterSettings.disableTeadsAudioSessionManagement()
            result(try? adapterSettings.asDictionary())
        case "userConsent":
            if let args = call.arguments as? [Any],
               let subjectToGDPR = args[0] as? String,
               let consent = args[1] as? String,
               let tcfVersionRawValue = args[2] as? Int,
               let tcfVersion = TCFVersion(rawValue: tcfVersionRawValue),
               let cmpSdkID = args[3] as? Int {
                adapterSettings.userConsent(subjectToGDPR: subjectToGDPR, consent: consent, tcfVersion: tcfVersion, cmpSdkID: cmpSdkID)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "setUsPrivacy":
            if let args = call.arguments as? [Any],
               let consent = args[0] as? String {
                adapterSettings.setUsPrivacy(consent: consent)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "disableBatteryMonitoring":
            adapterSettings.disableBatteryMonitoring()
            result(try? adapterSettings.asDictionary())
        case "addExtras":
            if let args = call.arguments as? [Any],
               let value = args[0] as? String,
               let key = args[1] as? String {
                adapterSettings.addExtras(value, for: key)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "enableDebug":
            adapterSettings.enableDebug()
            result(try? adapterSettings.asDictionary())
        case "enableValidationMode":
            adapterSettings.enableValidationMode()
            result(try? adapterSettings.asDictionary())
        case "pageUrl":
            if let args = call.arguments as? [Any],
               let urlString = args[0] as? String {
                adapterSettings.pageUrl(urlString)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "setMediaScale":
            if let args = call.arguments as? [Any],
               let mediaScaleIndex = args[0] as? Int,
               let mediaScale = MediaScale(rawValue: mediaScaleIndex) {
                adapterSettings.setMediaScale(mediaScale)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "addPlacementExtra":
            // Replicate from Android
            if let args = call.arguments as? [Any],
               let value = args[0] as? String,
               let key = args[1] as? String {
                adapterSettings.adPlacementSettings.addExtras(value, for: key)
                result(try? adapterSettings.asDictionary())
            } else {
                result(FlutterError.badArguments)
            }
        case "enableLocation":
            // Android only
            result(try? adapterSettings.asDictionary())
        case "useLightEndScreen":
            // Android only
            result(try? adapterSettings.asDictionary())
        case "hideBrowserUrl":
            // Android only
            result(try? adapterSettings.asDictionary())
        case "toolBarBackgroundColor":
            // Android only
            result(try? adapterSettings.asDictionary())
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: packages/TeadsAdmobAdapter/android/src/main/kotlin/tv/teads/teadsadmobadapter/teads_admob_adapter/FLTTeadsAdapterSettings.kt
package tv.teads.teadsadmobadapter.teads_admob_adapter

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import tv.teads.sdk.AdPlacementExtraKey
import tv.teads.sdk.TeadsMediationSettings
import tv.teads.sdk.plugin.PluginType
import tv.teads.sdk.renderer.MediaScale
import tv.teads.sdk.utils.userConsent.TCFVersion
import tv.teads.teadssdkflutter.teads_sdk_flutter.PluginException
import tv.teads.teadssdkflutter.teads_sdk_flutter.error

class FLTTeadsAdapterSettings : MethodChannel.MethodCallHandler {

    companion object {
        val instance by lazy { FLTTeadsAdapterSettings() }
    }

    internal var mediationSettingsBuilder = TeadsMediationSettings.Builder()

    init {
        mediationSettingsBuilder.addExtra(AdPlacementExtraKey.PLUGIN, PluginType.FLUTTER)
        mediationSettingsBuilder.addExtra(AdPlacementExtraKey.PLUGIN_VERSION, "1.0.0")
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "disableCrashMonitoring" -> {
                result.success(mediationSettingsBuilder.disableCrashMonitoring().build().toMap())
            }
            "userConsent" -> {
                (call.arguments as List<*>).let { args ->
                    val subjectToGDPR = args[0] as? String
                    val consent = args[1] as? String
                    val tcfVersion: Int? = args[2] as? Int
                    val cmpSdkId = args[3] as? Int
                    if (subjectToGDPR != null && consent != null && tcfVersion != null && cmpSdkId != null)
                        result.success(
                            mediationSettingsBuilder.userConsent(
                                subjectToGDPR, consent,
                                TCFVersion.fromInt(tcfVersion), cmpSdkId
                            ).build().toMap()
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "setUsPrivacy" -> {
                (call.arguments as List<*>).let { args ->
                    val consent = args[0] as? String
                    if (consent != null)
                        result.success(
                            mediationSettingsBuilder.setUsPrivacy(consent).build().toMap()
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "addExtras" -> {
                (call.arguments as List<*>).let { args ->
                    val value = args[0] as? String
                    val key = args[1] as? String
                    if (value != null && key != null)
                        result.success(
                            mediationSettingsBuilder.addExtra(key, value).build().toMap()
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "enableDebug" -> {
                result.success(mediationSettingsBuilder.enableDebug().build().toMap())
            }
            "enableValidationMode" -> {
                result.success(mediationSettingsBuilder.enableValidationMode().build().toMap())
            }
            "pageUrl" -> {
                (call.arguments as List<*>).let { args ->
                    val url = args[0] as? String
                    if (url != null)
                        result.success(
                            result.success(mediationSettingsBuilder.pageSlotUrl(url).build().toMap())
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "setMediaScale" -> {
                (call.arguments as List<*>).let { args ->
                    val mediaScaleInt = args[0] as? Int
                    if (mediaScaleInt != null) {
                        val mediaScale =
                            if (mediaScaleInt == 0) MediaScale.CENTER_CROP else MediaScale.CENTER_INSIDE
                        result.success(
                            result.success(
                                mediationSettingsBuilder.setMediaScale(mediaScale).build().toMap()
                            )
                        )
                    } else result.error(PluginException.BadArguments)
                }
            }
            "addPlacementExtra" -> {
                (call.arguments as List<*>).let { args ->
                    val value = args[0] as? String
                    val key = args[1] as? String
                    if (value != null && key != null)
                        result.success(
                            mediationSettingsBuilder.addPlacementExtra(key, value).build().toMap()
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "enableLocation" -> {
                result.success(mediationSettingsBuilder.enableLocation().build().toMap())
            }
            "useLightEndScreen" -> {
                result.success(mediationSettingsBuilder.useLightEndScreen().build().toMap())
            }
            "hideBrowserUrl" -> {
                result.success(mediationSettingsBuilder.hideBrowserUrl().build().toMap())
            }
            "toolBarBackgroundColor" -> {
                (call.arguments as List<*>).let { args ->
                    val color = args[0] as? Int
                    if (color != null)
                        result.success(
                            mediationSettingsBuilder.toolBarBackgroundColor(color).build().toMap()
                        )
                    else result.error(PluginException.BadArguments)
                }
            }
            "disableTeadsAudioSessionManagement", "disableBatteryMonitoring" -> {
                // iOS Only
                result.success(mediationSettingsBuilder.build().toMap())
            }
            else -> result.notImplemented()
        }
    }
}
